### DISCUSSION TITLE
Are exported sandboxes supposed to work standalone? Is there a "turbo" run in leopard?

### DISCUSSION THREAD

**Discussion by sfederici** [leopard-js/leopard#154] [Q&A]
I successully translated bodi3's project https://scratch.mit.edu/projects/673890545/.

It works (but the audio is audible only after clicking the "Green Flag" button twice).

Then I clicked the "Export to ZIP" button and I tried running the index.html file locally. It works only if I run it in a local webserver when the internet is on. Can someone tell me why? Scratch projects can be run locally in the standalone editor even when the internet is off.

I successfully translated also the project https://scratch.mit.edu/projects/643355824. To have a good sync, it requires turbo. Is there a way of turbo-running Scratch projects in Leopard?

Thanks for your help

**Comment by sfederici** [comment]
Is the Leopard project still managed and supported?

**Reply by towerofnix** [reply]
I'm not sure if it's being actively maintained — of the original devs, I'm pretty sure @PullJosh (who also started Leopard) is the only one who developed and maintains/maintained the website itself. I pop into issues once in a while but have limited availability, so don't consider myself a maintainer :P

I'll have a gander at your questions though:

> Are exported sandboxes supposed to work standalone?

Leopard doesn't require access to any online files except the compiled library files for Leopard itself:

* `https://unpkg.com/leopard@^1/dist/index.esm.js`
* `https://unpkg.com/leopard@^1/dist/index.min.css`

These are included near the top of `index.js` and `index.html` respectively. You can replace them with locally downloaded copies in the same directory for full offline/standalone access.

**However, you will still need a local server.** Sorry about the inconvenience here! This is a limitation of browser engines in recent years — previously accessing any path under the same directory was permitted from `file://`-opened URLs. [This had security implications](https://www.mozilla.org/en-US/security/advisories/mfsa2019-21/#CVE-2019-11730) (kind of unsurprisingly), so today browsers error when loading ["local [files] with included local resources"](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSRequestNotHttp). I couldn't find any spec-related details on what counts as "including local resources", but I gather JavaScript coded with ES6 modules counts, since those can further import dependencies (on a syntax level at JavaScript loading time, before runtime) from other paths or domains.

> Is there a way of turbo-running Scratch projects in Leopard?

See discussion in issue #83. My comment from last year:

> Turbo mode basically works by skipping screen refreshes and running as much code as possible until hitting the interval of 1/30th of a second (or whatever the specified frame rate is; at which point the current state of the project is rendered). It has some other additional quirks, so an in-depth investigation is required to make sure we implement this correctly if it's a desired feature.

There isn't a built-in way to do this, unfortunately, and we haven't done the necessary testing and compatibility-verification to get this to work.

Currently your best bet is to set a very high refresh rate in your `new Project` call at the end of `index.js`. If you're interested in working with Leopard's own code, you can try getting rid of [this `setInterval` call](https://github.com/leopard-js/leopard/blob/ceb1d9857951e7132103aa7a8bc2ab240a3af02f/src/Project.js#L37-L40) and providing your own calls to `step` (ex. continually evaluating until passing a 30ms refresh rate timer). Or you could "hack" your `Project` instance and override its `step` function, ex:

```js
// Totally untested
const oldStep = project.step.bind(project);
project.step = function() {
  const stopDate = Date.now() + 30;
  while (Date.now() < stopDate) {
    oldStep();
  }
};
```

**Reply by sfederici** [reply]
> However, you will still need a local server

Thanks, I understand. I tried running on my local web server a further project that I converted from [https://scratch.mit.edu/projects/807603336/](url). It works fine on codesandbox.io (almost, I had to "bypass" two "whenIReceive" calls in sprite1 as, for some reason, they were not triggered) . However, when I run it locally, this time I just get a large "undefined" icon in the middle of the browser page. Is it something that you know it can happen if you miss something?

**Reply by sfederici** [reply]
As suggested by @PullJosh at https://github.com/leopard-js/leopard/issues/75 (sorry, I had forgot) I got good results by removing the "yield;" at the end of for/while loops. While removing some yields makes the projects stuck (the whole codesandbox.io page freezes) as a general rule of thumb it seems to me that al least all "yields" at the end of loop where there is already a "wait" can be safely removed. Using this strategy I got the https://scratch.mit.edu/projects/673890545/ animation running much faster, so that the audio and the animation is now correctly synchronized.

This said, making turbo work inside Leopard would be an excellent way of showing how much faster it can be compared to Scratch.

**Reply by PullJosh** [reply]
Removing the yields at the end of loops is equivalent to what Scratch does inside of a "run without screen refresh" custom block. (Because `yield` literally means "pause here and continue after the screen has refreshed".)

However, I have no idea what Scratch's turbo mode actually does. Is it also removing yield points? Running as much logic as possible before updating the screen just in time for 30fps? Something else? I'm not sure, but maybe someone else knows...?

**Comment by sfederici** [comment]
It seems that the problem is FireFox. I opened "https://localhost:8080/index.html" in Chrome and it works fine.
